### PR TITLE
refactor: get version dynamically

### PR DESCRIPTION
**Pull request by jsjoeio**
# Description 

This PR refactors the release process to dynamically get the release version from the workflow inputs. This means we don't have to open a PR into `main` that modifies the `package.json`. It also means we'll need to manually modify the `CHANGELOG` and the Helm chart after doing a release.

## Todos

- [x] remove global `VERSION`
- [x] set anywhere we need it
- [x] standardize input to always include v and then strip when not needed
- [x] update release docs

## Screenshot

image

## Testing Plan

- [x] test build flow  - published successfully to npm [here](https://www.npmjs.com/package/@coder/code-server-pr/v/0.0.0-5753-9c704e6bc864b0e7511226fab77a5424ecf31b4d), successful build CI here (todo link)
- [x] test release flow: run [draft release manually](https://github.com/coder/code-server/actions/runs/3397555429/jobs/5649824113), and double-check assets
- [x] test publish flow: copy changes to fork and test `publish` [there](https://github.com/jsjoeio/code-server/actions/runs/3397622429/jobs/5649948994). uses the correct input

### CODE REVIEW

**Comment by jsjoeio** [comment]
First test: draft workflow succeeds. And when you download an asset, it shows the `package.json` has been updated and the `product.json`:

image

image

However, the assets have the wrong version in their name 🤔

image

I'm not sure what the root cause of this problem is though.

**Comment by jsjoeio** [comment]
Okay figured out the issue. we get `VERSION` from the root `package.json`. I need to refactor to be a function `version()` which takes an optional json path to get the version from.

File: .github/workflows/publish.yaml [hunk 1]
```diff
@@ -24,21 +29,23 @@ jobs:
...
       - name: Download npm package from release artifacts
         uses: robinraju/release-downloader@v1.5
         with:
           repository: "coder/code-server"
-          tag: v${{ steps.version.outputs.version }}
+          tag: ${{ github.event.inputs.version || github.ref_name }}
           fileName: "package.tar.gz"
           out-file-path: "release-npm-package"
 
+      - name: Get and set VERSION
```
Comment (code-asher) [inline .github/workflows/publish.yaml:40]: For a second I forgot why we needed to do this.  :laughing:   We might want to leave a comment that this exists just so we can strip out the `v` (same elsewhere this is used).

File: .github/workflows/release.yaml [hunk 2]
```diff
@@ -265,8 +296,24 @@ jobs:
...
+      - name: Modify version
+        run: |
+          echo "Updating version in root package.json"
+          npm version --prefix release ${{ inputs.version || github.ref_name }}
+
+          echo "Updating version in lib/vscode/product.json"
+          tmp=$(mktemp) 
+          jq '.codeServerVersion = "${{ inputs.version || github.ref_name }}"' release/lib/vscode/product.json > "$tmp" && mv "$tmp" release/lib/vscode/product.json
+
+      # TODO@jsjoeio - will it throw error if it already exists?
```
Comment (code-asher) [inline .github/workflows/release.yaml:311]: I tested and it seems it will override so we are good here :+1:

File: ci/build/build-vscode.sh [hunk 3]
```diff
@@ -47,6 +47,7 @@ main() {
   # as it is a submodule.
   export VSCODE_DISTRO_COMMIT
   VSCODE_DISTRO_COMMIT=$(git rev-parse HEAD)
+  VERSION=$(jq -r .version ../../package.json)
```
Comment (code-asher) [inline ci/build/build-vscode.sh:50]: Will this not be set by the workflow?

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -265,8 +296,24 @@ jobs:
...
+        run: tar -xzf package.tar.gz
+
+      - name: Modify version
+        run: |
+          echo "Updating version in root package.json"
+          npm version --prefix release ${{ inputs.version || github.ref_name }}
+
+          echo "Updating version in lib/vscode/product.json"
+          tmp=$(mktemp) 
+          jq '.codeServerVersion = "${{ inputs.version || github.ref_name }}"' release/lib/vscode/product.json > "$tmp" && mv "$tmp" release/lib/vscode/product.json
```
Comment (code-asher) [inline .github/workflows/release.yaml:309]: I think we need to strip the `v` on these.

File: .github/workflows/publish.yaml [hunk 5]
```diff
@@ -147,19 +161,21 @@ jobs:
...
           repository: "coder/code-server"
-          tag: v${{ steps.version.outputs.version }}
+          tag: v${{ env.VERSION }}
           fileName: "*.deb"
           out-file-path: "release-packages"
 
       - name: Publish to Docker
         run: yarn publish:docker
         env:
+          VERSION: ${{ steps.version.outputs.version }}
```
[the remark this answers is not in the document]
Comment (jsjoeio) [reply .github/workflows/publish.yaml:180]: Oh yes! Nice catch.

File: .github/workflows/publish.yaml (hunk 1 above)
Comment (jsjoeio) [reply .github/workflows/publish.yaml:40]: Gooooood idea!

File: ci/build/build-vscode.sh (hunk 3 above)
Comment (jsjoeio) [reply ci/build/build-vscode.sh:50]: AFAICT, we only build Code as part of the "Build code-server" job, which only runs in `build.yaml` which runs automatically so we have to grab it from the `package.json` here.

Is that right? Is there a better way to do this?

File: .github/workflows/release.yaml (hunk 4 above)
Comment (jsjoeio) [reply .github/workflows/release.yaml:309]: Nice catch!

File: ci/build/build-vscode.sh (hunk 3 above)
Comment (code-asher) [reply ci/build/build-vscode.sh:50]: Ahhh I saw `VERSION: "0.0.0"` and assumed that was for the build job but it was for the npm publish job.  I would be down to set `VERSION: "0.0.0"` for the build job as well.  That way the version is always set the same way (always by env var from a workflow).

**Comment by code-asher** [comment]
Should we also change the `package.json` version to be `development` or `0.0.0` or something like that?

File: ci/build/build-vscode.sh (hunk 3 above)
Comment (jsjoeio) [reply ci/build/build-vscode.sh:50]: Good idea. Okay I'll do `VERSION: "0.0.0"`

**Comment by jsjoeio** [comment]
> Should we also change the package.json version to be development or 0.0.0 or something like that?

Yeah, good call. I'll update it to `0.0.0`

**Comment by jsjoeio** [comment]
Ready for another review 👀

**Comment by jsjoeio** [comment]
e2e tests failing because it's using a cached version of Code. bumped cache key in secrets.

File: ci/build/build-vscode.sh (hunk 3 above)
Comment (code-asher) [inline ci/build/build-vscode.sh:50]: We still need to delete this line otherwise the version we set in the workflow gets overridden!

Comment (jsjoeio) [inline ci/build/build-vscode.sh:50]:
```suggestion
```

File: .github/workflows/publish.yaml [hunk 6]
```diff
@@ -73,6 +91,8 @@ jobs:
     timeout-minutes: 10
     env:
       GH_TOKEN: ${{ secrets.HOMEBREW_GITHUB_API_TOKEN }}
+      VERSION: ${{ github.event.inputs.version || github.ref_name }}
```
Comment (code-asher) [inline .github/workflows/publish.yaml:94]: Do we need this line since we set version in another step?

**Review by code-asher — APPROVED** [review]

File: .github/workflows/publish.yaml (hunk 6 above)
Comment (jsjoeio) [reply .github/workflows/publish.yaml:94]: I think so? Don't we need to set it in every step?

Comment (jsjoeio) [reply .github/workflows/publish.yaml:94]: Hmm double-checking..i see what you mean now

Comment (jsjoeio) [reply .github/workflows/publish.yaml:94]: Ah, you were right - nice catch. Removed!

**Pull request merged**